File: benchmarks/crown/ht/Output/Output/test_project/tests/test_ht.rs
use ntest::timeout;
use std::cell::RefCell;
use std::rc::Rc;
use test_project::ht::{
    hash_key, ht_create, ht_destroy, ht_expand, ht_get, ht_iterator, ht_length, ht_next, ht_set,
    ht_set_entry, Ht, HtEntry, HtIterator,
};
#[test]
#[timeout(60000)]
pub fn test_ht_update_value() {
    let mut table = ht_create::<i32>();

    // Insert a key-value pair
    let value1 = 42;
    let key = String::from("key");
    assert!(ht_set(&mut table, key.clone(), value1).is_some());

    // Update the value for the same key
    let value2 = 84;
    assert!(ht_set(&mut table, key.clone(), value2).is_some());

    // Retrieve the updated value
    let retrieved_value = ht_get(&table, key.clone());
    assert!(retrieved_value.is_some());
    assert_eq!(*retrieved_value.unwrap().borrow(), value2);

    // Check the length of the hash table (should still be 1)
    assert_eq!(ht_length(&table), 1);

    ht_destroy(table);
}

#[test]
#[timeout(60000)]
pub fn test_ht_memory_management() {
    let mut table = ht_create::<Rc<RefCell<i32>>>();

    let value1 = Rc::new(RefCell::new(42));
    let value2 = Rc::new(RefCell::new(84));

    assert!(ht_set(&mut table, "key1".to_string(), Rc::clone(&value1)).is_some());
    assert!(ht_set(&mut table, "key2".to_string(), Rc::clone(&value2)).is_some());

    let mut it = ht_iterator(Rc::new(RefCell::new(table)));
    loop {
        let (has_next, new_it) = ht_next(it);
        it = new_it;
        if !has_next {
            break;
        }
    }

    let table = Rc::try_unwrap(it._table).unwrap().into_inner();
    ht_destroy(table);
}

#[test]
#[timeout(60000)]
pub fn test_ht_set_and_get() {
    let mut table = ht_create::<i32>();

    // Insert a key-value pair
    let value1 = 42;
    let key1 = String::from("key1");
    assert!(ht_set(&mut table, key1.clone(), value1).is_some());

    // Retrieve the value
    let retrieved_value = ht_get(&table, key1.clone());
    assert!(retrieved_value.is_some());
    assert_eq!(*retrieved_value.unwrap().borrow(), value1);

    // Insert another key-value pair
    let value2 = 84;
    let key2 = String::from("key2");
    assert!(ht_set(&mut table, key2.clone(), value2).is_some());

    // Retrieve the second value
    let retrieved_value = ht_get(&table, key2.clone());
    assert!(retrieved_value.is_some());
    assert_eq!(*retrieved_value.unwrap().borrow(), value2);

    // Check the length of the hash table
    assert_eq!(ht_length(&table), 2);

    ht_destroy(table);
}

#[test]
#[timeout(60000)]
pub fn test_ht_create_and_destroy() {
    let mut table = ht_create::<i32>();
    assert_eq!(ht_length(&table), 0);
    ht_destroy(table);
}

#[test]
#[timeout(60000)]
pub fn test_ht_iterator() {
    let mut table = ht_create();
    let value1 = 1;
    let value2 = 2;
    let value3 = 3;
    assert!(ht_set(&mut table, "key1".to_string(), value1).is_some());
    assert!(ht_set(&mut table, "key2".to_string(), value2).is_some());
    assert!(ht_set(&mut table, "key3".to_string(), value3).is_some());

    let it = ht_iterator(Rc::new(RefCell::new(table)));
    let mut count = 0;
    let mut it = it;
    loop {
        let (has_next, new_it) = ht_next(it);
        it = new_it;
        if !has_next {
            break;
        }
        assert!(it.key.is_some());
        assert!(it.value.is_some());
        count += 1;
    }

    assert_eq!(count, 3);

    let table = it._table.borrow();
    let table_clone = Ht {
        entries: table.entries.clone(),
        capacity: table.capacity,
        length: table.length,
    };
    ht_destroy(table_clone);
}

File: benchmarks/crown/Input/test/test-ht.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "ht.h"

void test_ht_create_and_destroy() {
    ht* table = ht_create();
    assert(table != NULL);
    assert(ht_length(table) == 0);
    ht_destroy(table);
}

void test_ht_set_and_get() {
    ht* table = ht_create();
    assert(table != NULL);

    // Insert a key-value pair
    int value1 = 42;
    const char* key1 = "key1";
    assert(ht_set(table, key1, &value1) != NULL);

    // Retrieve the value
    int* retrieved_value = (int*)ht_get(table, key1);
    assert(retrieved_value != NULL);
    assert(*retrieved_value == value1);

    // Insert another key-value pair
    int value2 = 84;
    const char* key2 = "key2";
    assert(ht_set(table, key2, &value2) != NULL);

    // Retrieve the second value
    retrieved_value = (int*)ht_get(table, key2);
    assert(retrieved_value != NULL);
    assert(*retrieved_value == value2);

    // Check the length of the hash table
    assert(ht_length(table) == 2);

    ht_destroy(table);
}

void test_ht_update_value() {
    ht* table = ht_create();
    assert(table != NULL);

    // Insert a key-value pair
    int value1 = 42;
    const char* key = "key";
    assert(ht_set(table, key, &value1) != NULL);

    // Update the value for the same key
    int value2 = 84;
    assert(ht_set(table, key, &value2) != NULL);

    // Retrieve the updated value
    int* retrieved_value = (int*)ht_get(table, key);
    assert(retrieved_value != NULL);
    assert(*retrieved_value == value2);

    // Check the length of the hash table (should still be 1)
    assert(ht_length(table) == 1);

    ht_destroy(table);
}

void test_ht_iterator() {
    ht* table = ht_create();
    assert(table != NULL);

    // Insert multiple key-value pairs
    int value1 = 1, value2 = 2, value3 = 3;
    assert(ht_set(table, "key1", &value1) != NULL);
    assert(ht_set(table, "key2", &value2) != NULL);
    assert(ht_set(table, "key3", &value3) != NULL);

    // Iterate through the hash table
    hti it = ht_iterator(table);
    int count = 0;
    while (ht_next(&it)) {
        assert(it.key != NULL);
        assert(it.value != NULL);
        count++;
    }

    // Ensure all items were iterated
    assert(count == 3);

    ht_destroy(table);
}

void test_ht_memory_management() {
    ht* table = ht_create();
    assert(table != NULL);

    // Dynamically allocate memory for values
    int* value1 = malloc(sizeof(int));
    int* value2 = malloc(sizeof(int));
    *value1 = 42;
    *value2 = 84;

    assert(ht_set(table, "key1", value1) != NULL);
    assert(ht_set(table, "key2", value2) != NULL);

    // Free values during iteration
    hti it = ht_iterator(table);
    while (ht_next(&it)) {
        free(it.value);
    }

    ht_destroy(table);
}

int main() {
    test_ht_create_and_destroy();
    test_ht_set_and_get();
    test_ht_update_value();
    test_ht_iterator();
    test_ht_memory_management();

    printf("All tests passed!\n");
    return 0;
}